### ISSUE TITLE
Allow reading the actual float values from a GeoTIFF, using ImageReader.readRaster

### ISSUE THREAD

**Issue by thosmos** [haraldk/TwelveMonkeys#387]
I'm getting the following error when attempting to open a Float32 GeoTIFF file created by QGIS.  If I convert it first to Int16 then it works.  

```
reader:  #object[com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader 0x14933a00 com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader@14933a00]
#error {
 :cause Bad value for bits!
 :via
 [{:type java.lang.IllegalArgumentException
   :message Bad value for bits!
   :at [javax.imageio.ImageTypeSpecifier$Grayscale <init> ImageTypeSpecifier.java 584]}]
 :trace
 [[javax.imageio.ImageTypeSpecifier$Grayscale <init> ImageTypeSpecifier.java 584]
  [javax.imageio.ImageTypeSpecifier createGrayscale ImageTypeSpecifier.java 685]
  [com.twelvemonkeys.imageio.util.ImageTypeSpecifiers createGrayscale ImageTypeSpecifiers.java 111]
  [com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader getRawImageType TIFFImageReader.java 370]
  [com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader getImageTypes TIFFImageReader.java 730]
  [com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader read TIFFImageReader.java 759]
  [javax.imageio.ImageIO read ImageIO.java 1448]
  [javax.imageio.ImageIO read ImageIO.java 1308]
  [hellogeotiff$_main$fn__1390 invoke hellogeotiff.clj 104]
  [hellogeotiff$_main doInvoke hellogeotiff.clj 103]
  [clojure.lang.RestFn invoke RestFn.java 436]
  [hellogeotiff$eval1422 invoke form-init49144698386539134.clj 1]
  [clojure.lang.Compiler eval Compiler.java 6782]
  [clojure.lang.Compiler eval Compiler.java 6745]
  [clojure.core$eval invoke core.clj 3081]
  [clojure.main$repl$read_eval_print__7099$fn__7102 invoke main.clj 240]
  [clojure.main$repl$read_eval_print__7099 invoke main.clj 240]
  [clojure.main$repl$fn__7108 invoke main.clj 258]
  [clojure.main$repl doInvoke main.clj 258]
  [clojure.lang.RestFn invoke RestFn.java 1523]
  [clojure.tools.nrepl.middleware.interruptible_eval$evaluate$fn__706 invoke interruptible_eval.clj 87]
  [clojure.lang.AFn applyToHelper AFn.java 152]
  [clojure.lang.AFn applyTo AFn.java 144]
  [clojure.core$apply invoke core.clj 630]
  [clojure.core$with_bindings_STAR_ doInvoke core.clj 1868]
  [clojure.lang.RestFn invoke RestFn.java 425]
  [clojure.tools.nrepl.middleware.interruptible_eval$evaluate invoke interruptible_eval.clj 85]
  [clojure.tools.nrepl.middleware.interruptible_eval$interruptible_eval$fn__751$fn__754 invoke interruptible_eval.clj 222]
  [clojure.tools.nrepl.middleware.interruptible_eval$run_next$fn__746 invoke interruptible_eval.clj 190]
  [clojure.lang.AFn run AFn.java 22]
  [java.util.concurrent.ThreadPoolExecutor runWorker ThreadPoolExecutor.java 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run ThreadPoolExecutor.java 617]
  [java.lang.Thread run Thread.java 745]]}
```

**Comment by haraldk** [comment]
Hi Thomas,

I believe this is a duplicate of #374. 
Can you try to build the latest master, and see if that fixes the exception?

-- 
Harald K

**Comment by thosmos** [comment]
Ah, that looks promising.  When I tried to do `mvn package`, I got the following error:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.twelvemonkeys.lang.BeanUtilTestCase
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 sec - in com.twelvemonkeys.lang.BeanUtilTestCase
Running com.twelvemonkeys.lang.DateUtilTest
Tests run: 30, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.022 sec - in com.twelvemonkeys.lang.DateUtilTest
Running com.twelvemonkeys.lang.PlatformTest
Tests run: 14, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 0.006 sec - in com.twelvemonkeys.lang.PlatformTest
Running com.twelvemonkeys.lang.StringUtilTestCase
Tests run: 56, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.059 sec - in com.twelvemonkeys.lang.StringUtilTestCase
Running com.twelvemonkeys.lang.ValidateTest
Tests run: 114, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.044 sec - in com.twelvemonkeys.lang.ValidateTest
Running com.twelvemonkeys.util.BeanMapTestCase
Tests run: 35, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.078 sec - in com.twelvemonkeys.util.BeanMapTestCase
Running com.twelvemonkeys.util.CollectionUtilTest
Tests run: 34, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.07 sec - in com.twelvemonkeys.util.CollectionUtilTest
Running com.twelvemonkeys.util.convert.ConverterTestCase
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec - in com.twelvemonkeys.util.convert.ConverterTestCase
Running com.twelvemonkeys.util.convert.DateConverterTestCase
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec - in com.twelvemonkeys.util.convert.DateConverterTestCase
Running com.twelvemonkeys.util.convert.DefaultConverterTestCase
Tests run: 18, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.009 sec - in com.twelvemonkeys.util.convert.DefaultConverterTestCase
Running com.twelvemonkeys.util.convert.NumberConverterTestCase
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec - in com.twelvemonkeys.util.convert.NumberConverterTestCase
Running com.twelvemonkeys.util.convert.TimeConverterTestCase
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec - in com.twelvemonkeys.util.convert.TimeConverterTestCase
Running com.twelvemonkeys.util.LinkedMapTestCase
Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.046 sec - in com.twelvemonkeys.util.LinkedMapTestCase
Running com.twelvemonkeys.util.LRUMapTestCase
Tests run: 42, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.027 sec - in com.twelvemonkeys.util.LRUMapTestCase
Running com.twelvemonkeys.util.NullMapTestCase
Tests run: 35, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.007 sec - in com.twelvemonkeys.util.NullMapTestCase
Running com.twelvemonkeys.util.regex.RegExTokenIteratorTestCase
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 sec - in com.twelvemonkeys.util.regex.RegExTokenIteratorTestCase
Running com.twelvemonkeys.util.service.ServiceRegistryTest
Tests run: 25, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.031 sec - in com.twelvemonkeys.util.service.ServiceRegistryTest
Running com.twelvemonkeys.util.StringTokenIteratorTestCase
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec - in com.twelvemonkeys.util.StringTokenIteratorTestCase
Running com.twelvemonkeys.util.TimeoutMapTestCase
Tests run: 57, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.672 sec <<< FAILURE! - in com.twelvemonkeys.util.TimeoutMapTestCase
testWrapMapContainingValues(com.twelvemonkeys.util.TimeoutMapTestCase)  Time elapsed: 0.006 sec  <<< FAILURE!
junit.framework.AssertionFailedError: null
	at junit.framework.Assert.fail(Assert.java:47)
	at junit.framework.Assert.assertTrue(Assert.java:20)
	at junit.framework.Assert.assertNotNull(Assert.java:214)
	at junit.framework.Assert.assertNotNull(Assert.java:207)
	at com.twelvemonkeys.util.TimeoutMapTestCase.testWrapMapContainingValues(TimeoutMapTestCase.java:465)


Results :

Failed tests: 
  TimeoutMapTestCase.testWrapMapContainingValues:465 null

Tests run: 557, Failures: 1, Errors: 0, Skipped: 4

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Twelvemonkeys ...................................... SUCCESS [  5.920 s]
[INFO] TwelveMonkeys :: Common ............................ SUCCESS [  0.765 s]
[INFO] TwelveMonkeys :: Common :: Language support ........ FAILURE [  8.067 s]
[INFO] TwelveMonkeys :: Common :: IO ...................... SKIPPED
[INFO] TwelveMonkeys :: Common :: Image ................... SKIPPED
[INFO] TwelveMonkeys :: Servlet ........................... SKIPPED
[INFO] TwelveMonkeys :: ImageIO ........................... SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: Core ................... SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: Metadata ............... SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: Photoshop Path Support . SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: BMP plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: HDR plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: ICNS plugin ............ SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: IFF plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: JPEG plugin ............ SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: PCX plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: PDF plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: PICT plugin ............ SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: PNM plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: PSD plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: SGI plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: TGA plugin ............. SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: Thumbs.db plugin ....... SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: TIFF plugin ............ SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: Batik Plugin ........... SKIPPED
[INFO] TwelveMonkeys :: ImageIO :: reference test cases ... SKIPPED
[INFO] TwelveMonkeys :: Contrib ........................... SKIPPED
[INFO] TwelveMonkeys :: BOM ............................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 15.193 s
[INFO] Finished at: 2017-10-31T10:18:47-07:00
[INFO] Final Memory: 24M/246M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.18.1:test (default-test) on project common-lang: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/thomas/Develop/3d printer/terrain/TwelveMonkeys/common/common-lang/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :common-lang
```

**Comment by thosmos** [comment]
It built after I ran `mvn package` first in the `common` folder, then in the root folder.

**Comment by thosmos** [comment]
It is now passing the point where it was throwing the error before, but when I actually call .sampleRaster() the values are weird.  Would you like a sample image to see?  I switched to GDAL for now, but it would be good to have a more portable solution.

**Comment by haraldk** [comment]
Hi Thomas,

Yes,

Please add sample file + a minimal, standalone code sample (JUnit test case preferably) that shows the problem. 

`sampleRaster()` isn't a method I'm familiar with. What library/class is that? The values you get from the raster (either through the `BufferedImage` from `ImageReader.read(...)` or directly from `ImageReader.readAsRaster(...)` should be unnormalized 32 bit float values, directly as they are in the file. I'm not sure what you mean is "weird" about them..? :-)

-- 
Harald K

**Comment by haraldk** [comment]
Hmmm..

I just checked the code, and sorry, I see the float values are "clamped" to the range [0...1]... 

This is needed for proper display (as the Java2D color model expect that), however, that might not be what you need. The values should probably be normalized, instead of clamped too, depending on other TIFF tags 

I have a task, properly implementing `readRaster` to not do this clamping/normalization, and return the values exactly as is. 

But it would be nice to know what you would expect. :-)

-- 
Harald K

**Comment by thosmos** [comment]
Harald, thanks for looking into it.  Yes the problem I was seeing was that values that were expected to be in the 500.0 to 1700.0 range were clamped to 1.  However, -9999 made it through ... ???  

I'm doing this in Clojure [here](https://github.com/thosmos/GeoTiff2STL/blob/monkey/src/hellogeotiff.clj#L38) using the following calls:
```
^BufferedImage image (ImageIO/read file)
^Raster raster (.getData image)
sample (.getSample raster w h 0)
```
After taking a look at it a bit closer, it looks like `.getSample` returns an `int`, so I'm wondering how this corresponds to returning floats?  Will I need to cast the return of `.getData` to a different raster type?  Or will you expose a different function like `.getFloat`?

**Comment by haraldk** [comment]
Ah.. 

There's actually a bug in the clamping too. It only tests against the upper bound... :-P
Probably because the very few test images I have don't have negative values.

Anyway... I'm not fully up to speed on Clojure, but I think you should be able to just replace `getSample(x, y, b)` with `getSampleFloat(x, y, b)` to get the data as a `float`. Also, as the raster in this case should be backed by all floats, so using `getDataElements(x, y, null)` should work as well (and return an array of `float`s, with the length equal to the number of samples per pixel). The latter method is likely to be faster, as it requires less conversion (reusing the array parameter will make it faster still).


PS: I could really use a sample file + some way of testing how to get the expected samples...

-- 
Harald K

**Comment by thosmos** [comment]
OK, I've made a PR #389 that adds the sample file and a failing test for the getSampleFloat() call.

**Comment by haraldk** [comment]
Excellent!

I'll look into it.

-- 
Harald K

**Comment by converge20t** [comment]
I was having the same problem so I just checked out and compiled the latest code. Now when I run:

`BufferedImage image = ImageIO.read("greyscale_pic.tif");`

I don't get an exception anymore. However, whenever I try to access a black pixel using this code:

`int blackPixel = image.getRGB(x, y);`

I get an ArrayIndexOutOfBoundsException. I believe there is a problem with the TIFF reader code. I wrote a work-around and test for this exception but the code is not clean. Any ideas?

**Comment by haraldk** [comment]
@converge20t I don't see how this is related. 

Is your grayscale TIFF in float 32 format? Where is the stack trace?

Anyway, if you have a `BufferedImage`, the reader is done with its job... If you get an `AIOOBE` from the `getRGB` method, most likely you are using wrong values for `x` or `y`. But there's no way for me to verify this, as I don't have your image, and I don't know what your values are...

_If you still have this problem, please create a **new issue,** attach the TIFF that causes the problem and the minimal code to reproduce it._

Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
Okay, so here's the story... 

#394 and #374 are now fixed & closed. However, your PR still won't work. This is because the Java2D colour models expect the samples to be in range [0...1]. So to create a `BufferedImage` from float samples, we *need* to clamp the values to this range as we now do.

However, the `ImageIO` (`ImageReader`) API defines a separate method [`readRaster`](https://docs.oracle.com/javase/8/docs/api/javax/imageio/ImageReader.html#readRaster-int-javax.imageio.ImageReadParam-):

> Returns a new `Raster` object containing the raw pixel data from the image stream, without any color conversion applied. The application must determine how to interpret the pixel data by other means. Any destination or image-type parameters in the supplied `ImageReadParam` object are ignored, but all other parameters are used exactly as in the read method, except that any destination offset is used as a logical rather than a physical offset. The size of the returned `Raster` will always be that of the source region clipped to the actual image. Logical offsets in the stream itself are ignored.
>
>This method allows formats that normally apply a color conversion, such as JPEG, and formats that do not normally have an associated colorspace, such as remote sensing or medical imaging data, to provide access to raw pixel data.

I'm pretty sure this is what we want, and this method can safely return unnormalised floating point values, exactly as they are found in the TIFF stream.


Best regards,

-- 
Harald K

**Comment by thosmos** [comment]
OK, thanks for the info.  I may look into implementing readRaster() at some point as well.

**Comment by thosmos** [comment]
For anyone else facing this problem, I've found another probably more appropriate solution.  I've switched to using the GeoTools GeoTIFF reader.  This is part of the GeoTools full geospatial library, so it's probably better for someone using a GeoTIFF file anyway.  http://docs.geotools.org/stable/userguide/library/coverage/geotiff.html

**Comment by JimKhazan** [comment]
going through this too...
using readRaster does not seems to be the solution...

...
reader.setInput(input);
ImageReadParam param = reader.getDefaultReadParam();
param.setSourceRegion(new Rectangle(...));
r = reader.readRaster(0, param);

all values returned are clamped to 1.

So it seems there is no way to get the actual tiff values through TwelveMonkeys lib ?

**Comment by haraldk** [comment]
@JimKhazan No, *using* `readRaster` is currently not the solution at the moment, *implementing* it (the correct way), is. :-)

That's why this issue is still open. Feel free to have a look at though. All the code to read is there, you only need to modify it slightly to *not* normalize the values when reading the raster (but keep values normalized when reading an image, as this is a requirement of the API). 


Best regards,

-- 
Harald K

**Comment by bchapuis** [comment]
Can this issue be closed following the merging of #881?

**Comment by haraldk** [comment]
Fixed in #881
